src/librustc/middle/ty/relate.rs

```diff
@@ -191,14 +287,13 @@ fn relate_type_params<'a,'tcx:'a,R>(relation: &mut R,
                                                          &b_tys.len())));
     }
 
-    (0 .. a_tys.len())
+    relate_result_from_iter((0 .. a_tys.len())
         .map(|i| {
             let a_ty = a_tys[i];
             let b_ty = b_tys[i];
             let v = variances.map_or(ty::Invariant, |v| v[i]);
             relation.relate_with_variance(v, &a_ty, &b_ty)
-        })
-        .collect()
```

Comment:
if we replaced `relate_result_from_iter` with a collect impl, we could leave this the same... which would be nifty :)